### PR TITLE
**Review Only** Incorproate coreclr-debug into omnisharp extension

### PR DESCRIPTION
**Pull request by chuckries**
1. Adds coreclr-debug.ts to implement coreclr-debug acquisition and hooks
   it into omnisharp's activation handler
2. Adds the coreclr-debug directory containing the project.json,
   NuGet.config and dummy.cs required dotnet restore/publish
3. Adds the debugger section to contributes in package.json
4. Adds .vscodeignore to exclude items not needed at runtime from being
   packaged

TODO:
1. Update project.json references
2. Update NuGet.config to not point to http://dbgnuget (switch to
   nuget.org once pacakges are published publicly)
3. Create some completion file that we can check for correct complete
   installation
4. Improve reporting status of acquisition to end user

### CODE REVIEW

**Comment by chuckries** [comment]
I don't think this should be merged yet but I was hoping to get some feedback.

**Comment by chuckries** [comment]
@gregg-miskelly for discussion

**Comment by chuckries** [comment]
For the minimum supported VSCode version in the package.json, we would like to bump this up to whatever the next release of VSCode is. Do you have any issue with this?

File: src/coreclr-debug.ts [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        return '.exe'
+    }
+
+    return '';
+}
+
+function getPlatformLibExtension() : string {
+    switch (process.platform) {
+        case 'win32':
+            return '.lib';
```
Comment (chuckries) [inline src/coreclr-debug.ts]: this should be .dll

File: src/coreclr-debug.ts [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    
+    if (existsSync(_debugAdapterDir)) {
+        console.log('.NET Core Debugger tools already installed');
+        return;
+    }
+
+    _channel.appendLine("Downloading and configuring the .NET Core Debugger...");
+    _channel.show(vscode.ViewColumn.Three);
+    
+    vscode.window.setStatusBarMessage("Downloading and configuring the .NET Core Debugger...");
```
Comment (chuckries) [inline src/coreclr-debug.ts]: Remove.

File: package.json [hunk 3]
```diff
@@ -33,6 +33,7 @@
     },
     "activationEvents": [
         "onLanguage:csharp",
+        "onLanguage:fsharp",
```
Comment (DustinCampbell) [inline package.json]: Why is this needed?

File: package.json [hunk 4]
```diff
@@ -128,6 +129,228 @@
                 "language": "csharp",
                 "path": "./snippets/csharp.json"
             }
+        ],
+        "debuggers": [
+            {
+                "type": "coreclr",
+                "label": ".NET Core",
+                "enableBreakpointsFor": { "languageIds": [ "csharp", "fsharp" ] },
```
Comment (DustinCampbell) [inline package.json]: Note that the C# extension does not actually support editing of F# code. I'm not sure this is the right place for this.

File: src/coreclr-debug.ts [hunk 5]
```diff
@@ -0,0 +1,245 @@
+'use strict';
```
Comment (DustinCampbell) [inline src/coreclr-debug.ts]: Need copyright header. Look at other .ts files in the repo to get the boilerplate.

File: src/coreclr-debug.ts [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+import * as vscode from 'vscode';
+import * as child_process from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+
+var _coreClrDebugDir: string;
+var _debugAdapterDir: string;
+var _channel: vscode.OutputChannel;
```
Comment (DustinCampbell) [inline src/coreclr-debug.ts]: Best to use 'let' rather than 'var'. I'm working on moving things to ES6. Note that there's a tslint gulp task included that you can run in VS Code to find minor issues like this.

Comment (chuckries) [inline src/coreclr-debug.ts]: I'm fairly new to node/js. Do I need to use let throughout or just for these file level variables? My understanding is that let has block scope semantics vs global scope semantics?

Comment (DustinCampbell) [inline src/coreclr-debug.ts]: It's OK, I can change them later when I do a full sweep through the code base. Essentially, in ES6 it's "let" for variables that can be assigned to and "const" for variables that can't be assigned to.

Don't sweat it.

File: package.json [hunk 7]
```diff
@@ -128,6 +128,228 @@
...
+                },
+
+                "configurationAttributes": {
+                "launch": {
+                    "required": [ "program", "cwd" ],
+                    "properties": {
+                    "program": {
+                        "type": "string",
+                        "description": "Path to the program (executable file) to launch. On Windows, a '.exe' suffix is appended if not specified already.",
+                        "default": "${workspaceRoot}/bin/Debug/dnxcore50/<My-Project-Name>"
```
Comment (gregg-miskelly) [inline package.json]: We should merge in the most recent changes that @caslan made to this (removal of 'dnxcore50').

File: src/coreclr-debug.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    .then(function() {
+        return spawnChildProcess('dotnet', ['--verbose', 'publish', '-o', _debugAdapterDir], _channel, _coreClrDebugDir);
+    }).then(function() {
+        var promises: Promise<void>[] = [];
+
+        promises.push(renameDummyEntrypoint());
+        promises.push(removeLibCoreClrTraceProvider());
+
+        return Promise.all(promises);
+    }).then(function() {
```
Comment (gregg-miskelly) [inline src/coreclr-debug.ts]: Should we either check the version of 'dotnet' or check that it created the content file items to handle the case that the user has an older version of .NET CLI installed?

**Comment by DustinCampbell** [comment]
Closing in favor of #44.
